### PR TITLE
Add ws region

### PR DESCRIPTION
**Pull request by TheNeikos**
This first few commits will disappear once #279 is merged

Okay, so far I got this:
- [x] Send/Leave events
- [x] Cursor Updating
- [x] Fix hotspot setting

### CODE REVIEW

File: src/compositor/cursor.c [hunk 1]
```diff
@@ -104,6 +104,11 @@ ws_cursor_set_position(
     // off screen.
     self->x = CLAMP(-self->x_hp, x, w);
     self->y = CLAMP(-self->y_hp, y, h);
+    int retval = drmModeMoveCursor(self->cur_fb_dev->fd, self->cur_mon->crtc,
+            self->x, self->y);
+    if ( retval != 0) {
```
Comment (matthiasbeyer) [inline src/compositor/cursor.c]: Space after opening parentheses.

Comment (matthiasbeyer) [inline src/compositor/cursor.c]: Also, you could fix the indentation on the line above.

**Comment by TheNeikos** [comment]
Gotta love how clang probably just optimizes that away, while GCC gives a warning...

(It was complaining that `size_t` couldn't be negative, so checking for negativity seemed to be worthy of a warning... because to hell with generic programming!)

File: src/compositor/cursor.c [hunk 2]
```diff
@@ -91,6 +96,42 @@ ws_cursor_new(
...
+    if (real_x < x || real_x > (x + w)) {
+        return 0;
+    }
+
+    if (real_y < y || real_y > (y + h)) {
+        return 0;
+    }
+
+    if (ws_region_inside(surface->input_region, x - real_x, y - real_y)) {
+        *((struct ws_surface**) target) = (struct ws_surface*) _surface;
```
Comment (matthiasbeyer) [inline src/compositor/cursor.c:127]: What's the point of the cast on the left side of the assignment here?

Comment (matthiasbeyer) [inline src/compositor/cursor.c:127]: What's the point of the cast here anyways? `void` is explicitely casted, isn't it? Or am I getting something wrong here?

File: src/compositor/wayland/shell_surface.c [hunk 3]
```diff
@@ -221,8 +222,8 @@ ws_shell_surface_new(
 
     // try to set up the resource
     struct wl_resource* resource;
-    resource = wl_resource_create(client, &wl_shell_surface_interface,
-                                  WAYLAND_SHELL_SURFACE_VERSION, serial);
+    resource = ws_wayland_client_create_resource(client,
+            &wl_shell_surface_interface, WAYLAND_SHELL_SURFACE_VERSION, serial);
```
Comment (matthiasbeyer) [inline src/compositor/wayland/shell_surface.c:225]: Please add appropriate indentation.

File: src/input/input_device.c [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+ */
+
+static struct ws_logger_context log_ctx = { .prefix = "[Input/Module] " };
+
+static size_t
+hash_callback(
+    struct ws_object * _self
+);
+
+static int
```
Comment (matthiasbeyer) [inline src/input/input_device.c:53]: Please add at least a brief documentation for this and the function above.

File: src/compositor/cursor.c (hunk 2 above)
Comment (TheNeikos) [inline src/compositor/cursor.c:127]: You can't dereference a void pointer, so we cast it to a pointer pointer first

File: src/compositor/wayland/shell_surface.c (hunk 3 above)
Comment (TheNeikos) [inline src/compositor/wayland/shell_surface.c:225]: No, I am perfectly in sync with our guidelines + I'm at the hard limit of 80 characters.

Comment (matthiasbeyer) [inline src/compositor/wayland/shell_surface.c:225]: Jup, but the indentation for line 226 should be much more, to be in one column with the `client` argument.

**Comment by neithernut** [comment]
ACK from me so far.

**Comment by neithernut** [comment]
Targets #280, #281 and #285.

**Comment by neithernut** [comment]
I think you can squash, now.

**Comment by matthiasbeyer** [comment]
Yes, you can. But not all annotations are fixed by now.

File: src/compositor/wayland/shell_surface.c (hunk 3 above)
Comment (TheNeikos) [inline src/compositor/wayland/shell_surface.c:225]: So I should go over the 80 character limit? That makes no sense, and having a rotated L shaped function call just as much.

Comment (matthiasbeyer) [inline src/compositor/wayland/shell_surface.c:225]: You could put each argument on one line.

**Comment by matthiasbeyer** [comment]
Naaah, fuck it. ACK.

**Comment by neithernut** [comment]
ACK.

**Pull request merged**